### ISSUE TITLE
agnostic storage provider

### ISSUE THREAD

**Issue by gedw99** [fishworks/gofish#97]
I  am using this for package management and it occurred to me it would also be a nice addition to goFish

https://github.com/gomods/athens

It would allow GoFish to store the app binaries anywhere in any cloud but also onpremise.
It also seems to have a great velocity in the golang community.

https://github.com/gomods/athens/blob/master/pkg/repo/fetcher.go
Has a APi to github as GoFish will need it.

let me knwo what you think.

**Comment by bacongobbler** [comment]
What's the use case? Packages are fetched by a url. It does not matter where they are hosted.

**Comment by gedw99** [comment]
Hey @bacongobbler. 

true, however athens provides allot of the other bits in order to back it with different storage. As it is now GoFish does not provide that for you.

**Comment by bacongobbler** [comment]
Packages are defined and fetched from storage endpoints as URLs, but the storage backend is not an implementation detail of GoFish, so I think this request is irrelevant/out of scope. For example:

```
packages = {
    {
        os = "darwin",
        arch = "amd64",
        url = "https://github.com/golang/go/releases/download/v1.10.3/go-darwin-amd64",
        sha256 = "1544afdd4d543574ef8eabed343d683f7211202a65380f8b32035d07ce0c45ef",
        resources = {
            {
                path = name .. "-darwin-amd64",
                installpath = "bin/" .. name,
                executable = true
            }
        }
    }
}
```

Notice that the package defines a URL, but has no concept of who or what is hosting that package. it could be a blob store like Azure Blob Store or S3, or it could be a storage-agnostic webserver like minio or athens. GoFish doesn't care. GoFish cares only about how to fetch that package from a public URL.

In other words, the storage backend of the webserver hosting that package is not an implementation detail of the package manager but an implementation of the storage server (which GoFish does not provide).

Does that make sense?

**Comment by gedw99** [comment]
It Def makes sense. Thanks for your efforts to explain.

I will integrate Girish with Athens myself as I need it

I am doing an enterprise internal updating system.

I plan to deploy the update on all machines and use it with update all
onpremise apps with it.

Planning on making a taskbar GUI so users can control what they want to
update

Hopefully I can help but fix / PR.

**Comment by arschles** [comment]
Ref https://github.com/gomods/athens/issues/242

**Comment by bacongobbler** [comment]
closing as out of scope for gofish. Gofish is intended to be a package manager like apt, brew, pacman. Athens is a source code package repository, which is intended for an entirely different set of use cases.

As far as the original title (making gofish storage-agnostic), that's already part of the design since it fetches packages from a URL, therefore one just needs to host packages at a given URL.
